src/mb_caa_dimensions/DisplayedImage.tsx

```diff
@@ -0,0 +1,229 @@
+        } catch (e) {
+            LOGGER.error('Failed to load image information', e);
+            this.displayInfo('failed :(');
+        }
+    }
+
+    protected displayInfo(dimensionsString: string, fileInfoString?: string): void {
+        this.imgElement.setAttribute('ROpdebee_lazyDimensions', dimensionsString);
+
+        this.dimensionsSpan.textContent = dimensionsString;
+        if (typeof fileInfoString !== 'undefined') {
+            this.fileInfoSpan.textContent = fileInfoString;
+        }
+    }
+
+    protected createDimensionsString(imageInfo: ImageInfo): string {
+        return `Dimensions: ${createDimensionsString(imageInfo)}`;
+    }
+
+    protected createFileInfoString(imageInfo: ImageInfo): string | undefined {
+        const detailsString = createFileInfoString(imageInfo);
+        if (detailsString) {
+            return detailsString;
+        }
+        return undefined;
+    }
+}
+
+/**
+ * CAA images contained within an anchor element with `artwork-image` class.
+ *
+ * The full-size URL is the `href` of that anchor.
+ */
+export class ArtworkImageAnchorCAAImage extends DisplayedCAAImage {
+
+    constructor(imgElement: HTMLImageElement, cache: InfoCache) {
+        const fullSizeUrl = imgElement.closest<HTMLAnchorElement>('a.artwork-image')?.href;
+        assertDefined(fullSizeUrl);
+        super(imgElement, new CAAImage(fullSizeUrl, cache, imgElement.src));
+    }
+
+}
+
+/**
+ * CAA images on the cover art tab.
+ *
+ * Full-size URL needs to be retrieved from the anchors below the image.
+ */
+export class CoverArtTabCAAImage extends DisplayedCAAImage {
+
+    constructor(imgElement: HTMLImageElement, cache: InfoCache) {
+        const container = imgElement.closest('div.artwork-cont');
+        assertNonNull(container);
+        const fullSizeUrl = qs<HTMLAnchorElement>('p.small > a:last-of-type', container).href;
+        super(imgElement, new CAAImage(fullSizeUrl, cache));
+    }
+
+}
+
+/**
+ * CAA images with a `fullSizeURL` property.
+ *
+ * Intended for backward compatibility with other scripts.
+ */
+export class CAAImageWithFullSizeURL extends DisplayedCAAImage {
+    constructor(imgElement: HTMLImageElement, cache: InfoCache) {
+        const fullSizeUrl = imgElement.getAttribute('fullSizeURL');
+        assertNonNull(fullSizeUrl);
+        super(imgElement, new CAAImage(fullSizeUrl, cache));
+    }
+}
+
+/**
+ * Like `ArtworkImageAnchorCAAImage`, but shorter info string.
+ */
+export class ThumbnailCAAImage extends ArtworkImageAnchorCAAImage {
+
+    protected override createDimensionsString(imageInfo: ImageInfo): string {
+        return createDimensionsString(imageInfo);
+    }
+}
+
+export class DisplayedQueuedUploadImage extends BaseDisplayedImage {
+    private readonly image: QueuedUploadImage;
+
+    // No cache, unnecessary to cache.
+    constructor(imgElement: HTMLImageElement) {
+        super(imgElement);
+        this.image = new QueuedUploadImage(imgElement);
+
+    }
```

Comment:
```suggestion
    }
```
Spacing around the first and last members of a class is also a bit inconsistent right now (sometimes there is an empty line, sometimes not), but that doesn't itch me as much as this unexpected empty line 😝